### ISSUE TITLE
PLEASE make more requests run in parallel.

### ISSUE THREAD

**Issue by Pangamma** [handihow/skilltreev2#3]
**Is your feature request related to a problem? Please describe.**
This tool becomes unusable as more nodes are added. The reason is because requests for node info are made one by one, sequentially instead of in parallel. Each request takes roughly 200ms. I've actually experienced load times of over 1 minute due to this and am having to scan the internet for more efficient alternatives. There aren't many. So please optimize what we already have here. 

**Describe the solution you'd like**
- Fetch image data json in parallel or in bulk
- Fetch node data in parallel or in bulk.
- Reduce sequential fetches for data which will be poorly affected by latency times. 

**Describe alternatives you've considered**
I've tried to get the source code running locally, but am not currently familiar with firebase and do not know what effort would be required to set up the backend. 

**Additional context**
Try the following code fix: 
https://github.com/handihow/skilltreev2/blob/master/src/widgets/SkillTreeWidget.tsx#L35 

```
  const transformSkills = async (snap: QuerySnapshot) => {
        const skills: ISkill[] = await Promise.all(snap.docs.map(async doc => {
            const skill: ISkill = {
                parent: doc.ref.path.split("/"),
                path: doc.ref.path,
                ...(doc.data() as ISkill),
            };

            if (skill.image) {
                const resized = skill.image.split(".")[0] + "_128x128." + skill.image.split(".")[1];
                const reference = ref(storage, resized);
                skill.icon = await getDownloadURL(reference);
            }

            return skill;
        }));

        return skills;
    }
```

Anywhere you see the google chrome networking tab waterfall graph look like this (sequential) it is a good opportunity for optimization. 
![image](https://github.com/handihow/skilltreev2/assets/1046026/7aa50ae7-ff7f-4e78-9eb2-76852fd74dc0)

Basically, instead of doing 
```
const items = [];
for (const foo of array) {
const item = await getOperation();
items.push(item);
}
```
it is faster and better to do
``` 
const items = await Promise.all(array.map(async (const foo) => { return await getOperation(); }));
```

**Comment by handihow** [comment]
Thank you for taking the time to review the code and make this improvement request!

I have implemented the changes. Hope that this will lead to an improved experience.

**Comment by Pangamma** [comment]
You are awesome. Pretty sure that shaved 40s off the request time.

**Comment by handihow** [comment]
Yes sir!

Op ma 4 dec 2023 17:30 schreef Taylor Love ***@***.***>:
